libraries/AP_Avoidance/AP_Avoidance.cpp

```diff
@@ -0,0 +1,556 @@
+            }
+        }
+        should_change_state = true;
+    } while(0);
+
+    if (should_change_state) {
+
+        GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_WARNING, "AVOID: Moving from state (%s) to state (%s)", _state_names[_old_state], _state_names[new_state]);
+
+        _last_state_change_ms = now;
+
+        // we can change state but not change handler - for example,
+        // the user has specified the recovery handler for Fail->Warn
+        // as "continue fail behaviour".
+        bool change_handler = true;
+
+        if (_old_state == STATE_FAIL && new_state == STATE_WARN) {
+            // transitioning from fail -> warn
+            switch(_recovery_f) {
+            case AVOIDANCE_RECOVERY_F_CONTINUE_FAIL:
+                change_handler = false;
+                break;
+            case AVOIDANCE_RECOVERY_F_MOVE_TO_WARN:
+                change_handler = true;
+                break;
+            }
+        } else if (new_state == STATE_CLEAR) {
+            // transitioning from (fail or warn) to clear
+            switch(_recovery_w) {
+            case AVOIDANCE_RECOVERY_W_CONTINUE_ACTION:
+                change_handler = true;
+                break;
+            case AVOIDANCE_RECOVERY_W_RESUME:
+                change_handler = true;
+                break;
+            }
+        }
+
+        if (change_handler) {
+            AvoidanceHandler *new_avoidance_handler;
+            if (new_state == STATE_CLEAR) {
+                new_avoidance_handler = nullptr;
+            } else {
+                AvoidanceHandler &tmp = handler_for_action((MAV_COLLISION_ACTION)action);
+                new_avoidance_handler = &tmp;
+            }
+
+            if (new_avoidance_handler != _current_avoidance_handler) {
+                GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_WARNING, "AVOID: Moving to avoidance handler (%s)", ((new_avoidance_handler != nullptr) ? new_avoidance_handler->name() : "NULL"));
+                if (new_avoidance_handler == nullptr) {
+                    if (_current_avoidance_handler != nullptr) {
+                        _current_avoidance_handler->exit();
+                        _current_avoidance_handler = nullptr;
+                    }
+                } else {
+                    if (new_avoidance_handler->enter(*threat, _current_avoidance_handler)) {
+                        _current_avoidance_handler = new_avoidance_handler;
+                    } else {
+                        // whinge. This may not be an internal error -
+                        // e.g. we may have lost GPS
+                    }
+                }
+            }
+        }
+        _old_state = new_state;
+    }
+
+    if (_current_avoidance_handler != nullptr) {
+        _current_avoidance_handler->update();
+    }
+}
+
+
+
+// thanks to grumdrig (http://stackoverflow.com/questions/849211/shortest-distance-between-a-point-and-a-line-segment)
+float length_squared(const Vector2f v, const Vector2f w)
```

Comment:
@rmackay9 Possibly, but nobody else seems to have needed these.  I implemented these to not change the shape of the code borrowed from StackOverflow.

ATM the only static method in there is the one I added - "perpendicular", which is in Math simply because it was going to be a pain to write TESTs for it otherwise :-)